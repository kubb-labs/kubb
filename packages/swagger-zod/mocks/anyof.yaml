info:
  title: anyof test cases
  version: 1.0.0
openapi: 3.1.0
paths: {}
components:
  schemas:
    test:
      anyOf:
        - type: object
          properties:
            propertyA:
              type: string
          required:
            - propertyA
        - type: object
          properties:
            propertyA:
              type: string
            propertyB:
              type: string
          required:
            - propertyA
            - propertyB
