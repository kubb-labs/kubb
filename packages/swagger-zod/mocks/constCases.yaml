info:
  title: const test cases
  version: 1.0.0
openapi: 3.1.0
paths: {}
components:
  schemas:
    UuidSchema:
      type: string
      format: uuid
    NullableString:
      type:
        - string
        - "null"
    NullableStringWithAnyOf:
      anyOf:
        - type: string
        - type: "null"
    NullableStringUuid:
      type:
        - string
        - "null"
      format: uuid
    NullConst:
      required: true
      const: null
      description: |
        The value of this keyword MAY be null.
        Use of this keyword is functionally equivalent to an "enum" (Section 6.1.2) with a single value.
        An instance validates successfully against this keyword if its value is equal to the value of the keyword.
    StringValueConst:
      description: its value is equal to the value of the keyword
      required:
        - foobar
      properties:
        foobar:
          const: "foobar"
    NumberValueConst:
      description: its value is equal to the value of the keyword
      required:
        - foobar
      properties:
        foobar:
          const: 42
    MixedValueTypeConst:
      description: This probably should fail miserably
      required:
      - foobar
      properties:
        foobar:
          const: "foobar"
          type: number
