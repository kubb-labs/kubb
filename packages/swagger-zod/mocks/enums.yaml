definitions:
  enumVarNames.Type:
    enum:
    - 0
    - 1
    type: integer
    x-enum-varnames:
    - Pending
    - Received
  enumNames.Type:
    enum:
    - 0
    - 1
    x-enumNames:
      - Pending
      - Received
swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumVarNames.Type'
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNames.Type'
