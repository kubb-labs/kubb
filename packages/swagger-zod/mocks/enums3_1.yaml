openapi: 3.1.0
info:
  title: test
  version: "1.0.0"

components:
  schemas:
    enumNullable:
      type:
       - "string"
       - "null"
      enum:
      - Pending
      - Received

paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNullable'
