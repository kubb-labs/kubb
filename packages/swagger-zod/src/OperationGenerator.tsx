import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationSchema } from './components/OperationSchema.tsx'

import type { KubbFile } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import { Operations } from './components/Operations.tsx'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[], _operationsByMethod: OperationsByMethod): OperationMethodResult<FileMeta> {
    const { pluginManager, oas, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    const templates = {
      operations: Operations.templates,
      ...this.options.templates,
    }

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        {templates.operations && <Operations.File templates={templates.operations} />}
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  async operation(operation: Operation, options: PluginOptions['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })
    root.render(
      <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Oas.Operation operation={operation}>
          <OperationSchema.File mode={mode} />
        </Oas.Operation>
      </Oas>,
      { meta: { pluginManager, plugin: { ...plugin, options } } },
    )

    return root.files
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }
  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
