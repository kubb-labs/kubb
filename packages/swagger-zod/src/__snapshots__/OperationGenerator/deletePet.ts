[
  {
    "baseName": "delete_pet-petid.ts",
    "env": undefined,
    "exports": [],
    "id": undefined,
    "imports": [
      {
        "isTypeOnly": undefined,
        "name": [
          "z",
        ],
        "path": "zod",
        "root": undefined,
      },
    ],
    "language": "typescript",
    "meta": {
      "name": "delete_pet-petid",
      "pluginKey": undefined,
      "tag": undefined,
    },
    "override": undefined,
    "path": "delete_pet-petid.ts",
    "source": "export const DeletePetPetidMutationResponse = z.any();",
  },
]},
]