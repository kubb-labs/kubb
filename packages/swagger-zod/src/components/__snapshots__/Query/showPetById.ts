export const ShowPetByIdPathParams = z.object({
  petId: z.string().describe('The id of the pet to retrieve'),
  testId: z.string().describe('The id of the pet to retrieve'),
})

/**
 * @description Expected response to a valid request
 */
export const ShowPetById200 = z.lazy(() => Pet)

/**
 * @description unexpected error
 */
export const ShowPetByIdError = z.lazy(() => Error)

/**
 * @description Expected response to a valid request
 */
export const ShowPetByIdQueryResponse = z.lazy(() => Pet)
