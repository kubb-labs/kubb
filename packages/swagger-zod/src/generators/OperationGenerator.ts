import { getRelativePath } from '@kubb/core/utils'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'

import { ZodBuilder } from '../builders/index.ts'

import type { KubbFile } from '@kubb/core'
import type { FileResolver, Operation, OperationSchemas, Resolver } from '@kubb/swagger'
import type { FileMeta, PluginOptions } from '../types.ts'

// eslint-disable-next-line @typescript-eslint/ban-types
type Options = {}

export class OperationGenerator extends Generator<Options, PluginOptions> {
  resolve(operation: Operation): Resolver {
    const { pluginManager, plugin } = this.context

    return resolve({
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin.key,
    })
  }

  async all(): Promise<KubbFile.File | null> {
    return null
  }

  async get(operation: Operation, schemas: OperationSchemas, _options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { pluginManager, mode, plugin, oas } = this.context

    const zod = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: zod.name, pluginKey: plugin.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new ZodBuilder({ fileResolver: mode === 'file' ? undefined : fileResolver, withJSDocs: true, resolveName: pluginManager.resolveName, oas })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    return {
      path: zod.path,
      baseName: zod.baseName,
      source,
      imports: [
        {
          name: ['z'],
          path: 'zod',
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async post(operation: Operation, schemas: OperationSchemas, _options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { pluginManager, mode, plugin, oas } = this.context

    const zod = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: zod.name, pluginKey: plugin.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new ZodBuilder({ fileResolver: mode === 'file' ? undefined : fileResolver, withJSDocs: true, resolveName: pluginManager.resolveName, oas })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.request)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    return {
      path: zod.path,
      baseName: zod.baseName,
      source,
      imports: [
        {
          name: ['z'],
          path: 'zod',
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async put(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async patch(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async delete(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
}
