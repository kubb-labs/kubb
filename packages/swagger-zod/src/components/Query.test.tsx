import { mockedPluginManager } from '@kubb/core/mocks'
import { createRootServer } from '@kubb/react/server'
import { OasManager } from '@kubb/swagger'

import { OperationGenerator } from '../OperationGenerator.tsx'
import { Query } from './Query.tsx'

import type { Plugin } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from '../types.ts'

describe('<Query/>', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-zod/mocks/petStore.yaml' },
  })

  const options: GetOperationGeneratorOptions<OperationGenerator> = {
    exclude: [],
    include: undefined,
    override: undefined,
    transformers: {},
    typed: false,
    dateType: 'string',
    unknownType: 'any',
  }

  const plugin = { options } as Plugin<PluginOptions>
  const og = await new OperationGenerator(
    options,
    {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
    },
  )

  test('showPetById', async () => {
    const operation = oas.operation('/pets/{petId}', 'get')
    const schemas = og.getSchemas(operation)
    const context: AppContextProps<PluginOptions['appMeta']> = { meta: { oas, pluginManager: mockedPluginManager, plugin, schemas, operation } }

    const Component = () => {
      return <Query.File mode="directory" />
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/Query/showPetById.ts')
  })
})
