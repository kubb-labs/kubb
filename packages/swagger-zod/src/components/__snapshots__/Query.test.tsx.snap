// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Query/> > showPetById 1`] = `
"export const ShowPetByIdPathParams = z.object({
  petId: z.string().describe(\`The id of the pet to retrieve\`),
  testId: z.string().describe(\`The id of the pet to retrieve\`),
})

/**
 * @description Expected response to a valid request
 */
export const ShowPetById200 = z.lazy(() => Pet)

/**
 * @description unexpected error
 */
export const ShowPetByIdError = z.lazy(() => Error)

/**
 * @description Expected response to a valid request
 */
export const ShowPetByIdQueryResponse = z.lazy(() => Pet)
"
`;
